src/Hamiltonians/Product.jl

```diff
@@ -0,0 +1,163 @@
+"""
+    OperatorProduct(A::AbstractOperator, B::AbstractOperator)
+
+The product of two [`AbstractOperator`](@ref)s, acting from right to left. The two operators
+must act on the same address space.
+"""
+struct OperatorProduct{T, O1<:AbstractOperator, O2<:AbstractOperator} <: AbstractOperator{T}
+    op1::O1
+    op2::O2
+end
+function OperatorProduct(op1::AbstractOperator{T1}, op2::AbstractOperator{T2}) where {T1, T2}
+    if (op2 isa AbstractHamiltonian && !allows_address_type(op1, starting_address(op2)) ||
+        op1 isa AbstractHamiltonian && !allows_address_type(op2, starting_address(op1)))
```

Comment:
`AbstractOperator`s don't necessarily have `starting_address` defined. You'd need `AbstractHamiltonian`s for this. Maybe you want a call signature like this:
```julia
function OperatorProduct(op1::AbstractOperator{T1}, op2::AbstractHamiltonian{T2}) where {T1, T2}
```
Another reason for this is that the `eltype` of `AbstractOperator` can be a matrix, which would create problems that you may want to avoid.